Source/Core/Core/PowerPC/Jit64/Jit64_Util.cpp

```diff
@@ -0,0 +1,420 @@
+  auto val =
+      (reg_value.IsImm() || reg_value.IsRegBound()) ? reg_value : reg_value.Bind(BindMode::Read);
+
+  // Optimize swap flag
+  if (swap)
+  {
+    // No need to swap
+    if (accessSize == 8)
+    {
+      swap = false;
+    }
+    else
+    {
+      // We can swap these ahead of time
+      if (val.IsImm())
+      {
+        swap = false;
+        switch (accessSize)
+        {
+        case 16:
+          val = regs.gpr.Imm32(Common::swap16(val.Imm32()));
+          break;
+        case 32:
+          val = regs.gpr.Imm32(Common::swap32(val.Imm32()));
+          break;
+        case 64:
+          // We don't support 64-bit immediates (TODO?)
+          UnexpectedInstructionForm();
+          break;
+        }
+      }
+    }
+  }
+  u32 flags = swap ? 0 : SAFE_LOADSTORE_NO_SWAP;
+
+  // We need to bind a register, so may as well just add the offset here
+  if (reg_addr.IsImm())
+  {
+    if (update)
+    {
+      // Update reg_addr inside a transaction
+      auto xa = reg_addr.Bind(BindMode::ReadWriteTransactional);
+      MOV_sum(32, xa, xa, offset);
+      SafeWriteRegToReg(val, xa, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    }
+    else
+    {
+      auto scratch = regs.gpr.Borrow();
+      MOV_sum(32, scratch, reg_addr, offset);
+      SafeWriteRegToReg(val, scratch, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    }
+    return;
+  }
+
+  auto addr = reg_addr.Bind(BindMode::Read);
+
+  // Easy: zero offset, so updates don't matter
+  if (offset.IsZero())
+  {
+    SafeWriteRegToReg(valueAddressShareRegister ? addr : val, addr, accessSize, 0,
+                      CallerSavedRegistersInUse(), flags);
+    return;
+  }
+
+  // If we're updating we can clobber address (transactionally)
+  if (update)
+  {
+    if (valueAddressShareRegister)
+    {
+      // Keep value in a scratch register so we can update the address
+      // without worrying about it
+      auto scratch = reg_value.BorrowCopy();
+      auto xa = addr.Bind(BindMode::ReadWriteTransactional);
+      MOV_sum(32, xa, xa, offset);
+      SafeWriteRegToReg((X64Reg)scratch, xa, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    }
+    else
+    {
+      // Update reg_addr inside a transaction
+      auto xa = addr.Bind(BindMode::ReadWriteTransactional);
+      MOV_sum(32, xa, xa, offset);
+      SafeWriteRegToReg(val, xa, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    }
+    return;
+  }
+
+  if (valueAddressShareRegister)
+  {
+    // Not transactional, since we only modify value by writing to it
+    auto xv = reg_value.Bind(BindMode::ReadWrite);
+    auto scratch = regs.gpr.Borrow();
+    MOV_sum(32, scratch, xv, offset);
+    SafeWriteRegToReg((X64Reg)xv, scratch, accessSize, 0, CallerSavedRegistersInUse(), flags);
+    return;
+  }
+
+  if (valueOffsetShareRegister)
+  {
+    // Not transactional, since we only modify value by writing to it
+    auto xv = reg_value.Bind(BindMode::ReadWrite);
```

Comment:
If `WriteClobbersRegValue` we allocate a scratch register and recurse back into the same function, otherwise we fall back to the N cases below that test for the different combinations of constant/zero-offset, etc.